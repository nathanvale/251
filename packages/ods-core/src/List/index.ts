export { List } from "./List";
export type ListProps = import("./List").ListProps;
