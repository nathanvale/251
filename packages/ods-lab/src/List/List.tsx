import * as React from "react";
import MuiList, {ListProps as IList} from "@material-ui/core/List";

export interface ListProps extends IList {}

export const List = (props: ListProps) => (
  <MuiList {...props}>{props.children}</MuiList>
);
