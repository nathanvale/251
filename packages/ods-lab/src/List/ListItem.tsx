import * as React from "react";
import MuiListItem, {
  ListItemProps as IListItem,
} from "@material-ui/core/ListItem";

export interface ListItemProps extends IListItem {}

export const ListItem = (props: ListItemProps) => {
  const { button, ...other } = props;
  return <MuiListItem {...other} />;
};
