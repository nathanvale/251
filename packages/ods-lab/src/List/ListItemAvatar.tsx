import * as React from "react";
import MuiListItemAvatar, {
  ListItemAvatarProps as IListItemAvatar,
} from "@material-ui/core/ListItemAvatar";

export interface ListItemAvatarProps extends IListItemAvatar {}

export const ListItemAvatar = (props: ListItemAvatarProps) => (
  <MuiListItemAvatar {...props} />
);
