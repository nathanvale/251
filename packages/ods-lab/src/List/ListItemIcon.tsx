import * as React from "react";
import MuiListItemIcon, {
  ListItemIconProps as IListItemIcon,
} from "@material-ui/core/ListItemIcon";

export interface ListItemIconProps extends IListItemIcon {}

export const ListItemIcon = (props: ListItemIconProps) => (
  <MuiListItemIcon {...props}>{props.children}</MuiListItemIcon>
);
