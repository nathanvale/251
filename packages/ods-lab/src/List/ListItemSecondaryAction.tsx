import * as React from "react";
import MuiListItemSecondaryAction, {
  ListItemSecondaryActionProps as IListItemSecondaryAction,
} from "@material-ui/core/ListItemSecondaryAction";

export interface ListItemSecondaryActionProps
  extends IListItemSecondaryAction {}

export const ListItemSecondaryAction = (
  props: ListItemSecondaryActionProps,
) => (
  <MuiListItemSecondaryAction {...props}>
    {props.children}
  </MuiListItemSecondaryAction>
);
