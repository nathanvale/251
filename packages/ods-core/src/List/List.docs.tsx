import { ComponentDocs } from "@origin-digital/ods-types";
import { docs as listItemDocs } from "../ListItem/ListItem.docs";
import { docs as textDocs } from "../Text/Text.docs";
import { ListProps } from "./List";

export const docs: ComponentDocs<ListProps> = {
  category: "Content",
  componentName: "List",
  examples: listItemDocs.examples,
  description:
    "ODS offers authors several mechanisms for specifying lists of information. All lists must contain one or more list elements. Lists may contain unordered information, ordered information or icons.",
  propDescriptions: {
    space: "The responsive prop T-shirt sized space between all items.",
    tone: textDocs?.propDescriptions?.tone,
    variant: textDocs?.propDescriptions?.variant,
    weight: textDocs?.propDescriptions?.weight,
    type: "Defines the type of the list item marker (bullet).",
  },
  snippets: [],
};
