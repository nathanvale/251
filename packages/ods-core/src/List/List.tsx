import styled, { css } from "styled-components";
import React, { useMemo } from "react";
import {
  RequiredWithoutChildren,
  OptionalTrackableProps,
} from "@origin-digital/ods-types/";
import { style, get } from "styled-system";
import { cssLengthToString } from "@origin-digital/ods-helpers";
import { TextProps, useTextStyles } from "../Text/Text";
import { Stack, StackProps } from "../Stack/Stack";
import { CounterContainer, CounterLabel, Counter } from "../ListItem/ListItem";
import { TextContext } from "../Text/TextContext";

type ULType = "disc" | "circle";
type OLType =
  | "decimal"
  | "lower-alpha"
  | "upper-alpha"
  | "lower-roman"
  | "upper-roman"
  | "decimal-stylised";

const listDefaultProps: RequiredWithoutChildren<ListProps> = {
  variant: "body",
  space: ["small", "medium"],
  tone: "neutral",
  type: "disc",
  weight: "regular",
  "data-id": "list",
};

export interface ListProps extends OptionalTrackableProps {
  children: StackProps["children"];
  variant?: Exclude<TextProps["variant"], "caption" | "overline-text">;
  space?: StackProps["space"];
  tone?: TextProps["tone"];
  weight?: TextProps["weight"];
  type?: ULType | OLType;
}

const marginBlockStart = style({
  prop: "space",
  cssProperty: "margin-block-start",
  key: "space",
  transformValue: (n, scale) => cssLengthToString(get(scale, n)),
});

const getCounterStyles = ({
  type,
  variant,
  theme,
}: Pick<ListProps, "type" | "variant"> & { theme: any }) => {
  const isSmall = variant === "body-small" || variant === "subtitle-small";
  if (type === "decimal-stylised") {
    return css`
      border-radius: 50%;
      background: ${theme?.colors?.grey500};
      display: flex;
      align-items: center;
      justify-content: center;
      width: ${isSmall ? "20" : "24"}px;
      height: ${isSmall ? "20" : "24"}px;
    `;
  } else if (type === "disc" || type === "circle") {
    return css`
      width: ${isSmall ? "16" : "18"}px;
      height: ${isSmall ? "20" : "24"}px;
      justify-content: flex-end;
    `;
  } else {
    return css`
      width: ${isSmall ? "18" : "20"}px;
      height: ${isSmall ? "20" : "24"}px;
      justify-content: flex-end;
    `;
  }
};

const StylisedList = styled(Stack)<{
  type: ListProps["type"];
  space: StackProps["space"];
  variant: ListProps["variant"];
}>`
  &&& {
    margin-block-start: 0px;
    margin-block-end: 0px;
    ul {
      ${(p) => marginBlockStart(p)}
    }
    ol {
      ${(p) => marginBlockStart(p)}
    }
    margin-inline-start: 0px;
    margin-inline-end: 0px;
    padding-inline-start: 0px;
    list-style: none;
    counter-reset: listCounter;
    li ${CounterContainer} {
      min-width: ${(p) =>
        p.variant === "body-small" || p.variant === "subtitle-small"
          ? "28"
          : "32"}px;
      height: ${(p) =>
        p.variant === "body-small" || p.variant === "subtitle-small"
          ? "20"
          : "24"}px;
      display: flex;
      counter-increment: listCounter;
      align-items: flex-end;
      ${Counter} {
      border-radius: none;
      background: none;
      display: flex;
      align-items: flex-start;
        ${(p) =>
          getCounterStyles({
            type: p.type,
            variant: p.variant,
            theme: p.theme,
          })}
        ${CounterLabel} {
          color: inherit;
          ${(p) =>
            p.type === "decimal-stylised"
              ? css`
                  color: ${p.theme.colors.white};
                  line-height: 24px;
                  font-size: 14px;
                `
              : ""}
        }
      }
      ${CounterLabel} {
        &::before {
          content: counter(
              listCounter,
              ${(p) => (p.type === "decimal-stylised" ? "decimal" : p.type)}
            )
            "${(p) =>
              p.type === "disc" ||
              p.type === "circle" ||
              p.type === "decimal-stylised"
                ? ""
                : "."}";
        }
      }
    }
  }
`;

export const List = ({
  children,
  variant = listDefaultProps.variant,
  space = listDefaultProps.space,
  tone = listDefaultProps.tone,
  type = listDefaultProps.type,
  weight = listDefaultProps.weight,
  "data-id": dataId,
}: ListProps) => {
  // Prevent re-renders when context values haven't changed
  const textContextValue = useMemo(
    () => ({
      tone,
      weight,
      variant,
    }),
    [tone, weight, variant]
  );
  const className = useTextStyles({ weight, tone, variant });
  const component = type === "circle" || type === "disc" ? "ul" : "ol";
  return (
    <TextContext.Provider value={textContextValue}>
      <StylisedList
        className={className}
        component={component}
        data-id={dataId}
        space={space}
        variant={variant}
        type={type}
      >
        {children}
      </StylisedList>
    </TextContext.Provider>
  );
};

List.defaultProps = listDefaultProps;
List.displayName = "List";
