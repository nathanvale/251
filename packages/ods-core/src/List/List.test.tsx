import React from "react";
import { render } from "@origin-digital/ods-testing-library";
import { List } from "./List";

test("it renders a List", () => {
  const { container } = render(
    <List>
      <List>This is a secondary ListItem.</List>
      <List>This is a secondary ListItem.</List>
      <List>This is a secondary ListItem.</List>
    </List>
  );
  expect(container.firstChild).toMatchSnapshot();
});

test("it renders a List with custom data-id", () => {
  const { container } = render(
    <List data-id="myId">
      <List>This is a secondary ListItem.</List>
      <List>This is a secondary ListItem.</List>
      <List>This is a secondary ListItem.</List>
    </List>
  );

  expect(container.firstChild).toHaveAttribute("data-id", "myId");
});
