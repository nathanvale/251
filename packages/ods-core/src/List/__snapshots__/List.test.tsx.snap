// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a List 1`] = `
.c3 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  display: list-item;
}

.c1 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  display: block;
  width: 100%;
}

.c2:empty {
  display: none;
}

.c2:empty + div:before {
  display: none;
}

.c2:not(:empty) ~ div:not(:empty):before {
  content: "";
  display: block;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  position: relative;
  height: 0.1px;
  margin-top: 12px;
}

.c2:nth-child(n + 2):before {
  content: "";
  display: block;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  position: relative;
  height: 0.1px;
  margin-top: 12px;
}

.c0.c0.c0 {
  margin-block-start: 0px;
  margin-block-end: 0px;
  margin-inline-start: 0px;
  margin-inline-end: 0px;
  padding-inline-start: 0px;
  list-style: none;
  counter-reset: listCounter;
}

.c0.c0.c0 ul {
  margin-block-start: 12px;
}

.c0.c0.c0 ol {
  margin-block-start: 12px;
}

.c0.c0.c0 li .c4 {
  min-width: 32px;
  height: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  counter-increment: listCounter;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c0.c0.c0 li .c4 .c5 {
  border-radius: none;
  background: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  width: 18px;
  height: 24px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0.c0.c0 li .c4 .c5 .c6 {
  color: inherit;
}

.c0.c0.c0 li .c4 .c6::before {
  content: counter( listCounter,disc ) "";
}

@media screen and (min-width:576px) {
  .c2:not(:empty) ~ div:not(:empty):before {
    margin-top: 12px;
  }
}

@media screen and (min-width:768px) {
  .c2:not(:empty) ~ div:not(:empty):before {
    margin-top: 12px;
  }
}

@media screen and (min-width:992px) {
  .c2:not(:empty) ~ div:not(:empty):before {
    margin-top: 16px;
  }
}

@media screen and (min-width:1200px) {
  .c2:not(:empty) ~ div:not(:empty):before {
    margin-top: 16px;
  }
}

@media screen and (min-width:576px) {
  .c2:nth-child(n + 2):before {
    margin-top: 12px;
  }
}

@media screen and (min-width:768px) {
  .c2:nth-child(n + 2):before {
    margin-top: 12px;
  }
}

@media screen and (min-width:992px) {
  .c2:nth-child(n + 2):before {
    margin-top: 16px;
  }
}

@media screen and (min-width:1200px) {
  .c2:nth-child(n + 2):before {
    margin-top: 16px;
  }
}

@media screen and (min-width:576px) {
  .c0.c0.c0 ul {
    margin-block-start: 16px;
  }
}

@media screen and (min-width:576px) {
  .c0.c0.c0 ol {
    margin-block-start: 16px;
  }
}

<ul
  class="typography-body-variant-basekick-1 typography-neutral-color-2 typography-regular-weight-3 typography-strong-support-4 c0 -ul c1"
  data-id="list"
  display="block"
  width="full"
>
  <li
    class="c2 -li c3"
    data-id="stack-child"
    display="list-item"
  >
    <ul
      class="typography-body-variant-basekick-5 typography-neutral-color-6 typography-regular-weight-7 typography-strong-support-8 c0 -ul c1"
      data-id="list"
      display="block"
      width="full"
    >
      <li
        class="c2 -li c3"
        data-id="stack-child"
        display="list-item"
      >
        This is a secondary ListItem.
      </li>
    </ul>
  </li>
  <li
    class="c2 -li c3"
    data-id="stack-child"
    display="list-item"
  >
    <ul
      class="typography-body-variant-basekick-9 typography-neutral-color-10 typography-regular-weight-11 typography-strong-support-12 c0 -ul c1"
      data-id="list"
      display="block"
      width="full"
    >
      <li
        class="c2 -li c3"
        data-id="stack-child"
        display="list-item"
      >
        This is a secondary ListItem.
      </li>
    </ul>
  </li>
  <li
    class="c2 -li c3"
    data-id="stack-child"
    display="list-item"
  >
    <ul
      class="typography-body-variant-basekick-13 typography-neutral-color-14 typography-regular-weight-15 typography-strong-support-16 c0 -ul c1"
      data-id="list"
      display="block"
      width="full"
    >
      <li
        class="c2 -li c3"
        data-id="stack-child"
        display="list-item"
      >
        This is a secondary ListItem.
      </li>
    </ul>
  </li>
</ul>
`;
